man/man3/loadobj.3
.TH LOADOBJ 3NEMO "6 November 1991"
.SH NAME
loadobj, findfn, mysymbols \- dynamic object file loader
.SH SYNOPSIS
.nf
.B #include <stdinc.h>
.B #include <loadobj.h>
.PP
.B void loadobj(pathname)
.B string pathname
.PP
.B proc findfn(fname)
.B string fname;
.PP
.B void mysymbols(progname)
.B string progname;
.SH DESCRIPTION
These routines provide a uniform low-level I/O interface to loading
object modules. Higher level I/O interfaces are present for a few
standard operations (see e.g. \fIpotential(3NEMO)\fP and 
\fIbodytrans(3NEMO))\fP.
.PP
These routines are intended for applications where a set of mathematical
expressions is specified at runtime as function strings, which are then
evaluated for a large number of variable values. This is done by compiling the
set of function strings into byte code sequences, which can be interpreted very
efficiently for the various variable values. 
.PP
The routine \fImysymbols\fP is generally called once and in the beginning
of the program. It loads all symbols from the program pointed to
by the argument \fIprogname\fP, which is normally the current 
program. These symbols are then stored in an internally defined symbol tree
\fBsymtree\fP. In case \fIprogname\fP is not the complete name, the
\fBPATH\fP environment variable is used (see \fIpathopen(3NEMO)\fP).
.PP
The undefined symbols in each object module that is loaded with 
\fIloadobj\fP are resolved against the already existing symbol tree, 
and new symbols are added to the tree. If a symbol is not found to
be resolved, this is considered an error, and \fIerror(3NEMO)\fP is
called, which aborts the program. In addition the code loaded into
memory is properly relocated using information in the object
code. The argument \fIpathname\fP must be the proper filename, no
search path is provided for.
.PP
After all symbols have been defined, the programmer can obtain
pointers to functions which is specified by name using \fIfindfn\fP.
.SH EXAMPLE
In the following section of code gives an example of use.
.nf
    proc fn;                   /* see defs in <stdinc.h> */

    mysymbols(getparam("argv0"));   /* get local symbols */
    loadobj("test.o");               /* load object code */
    fn = (proc) findfn("_test");       /* find function  */
    if (fn==NULL) error("No test");       /* catch error */
    (*fn)();         /* execute routine test() in test.o */
.fi
.SH SEE ALSO
a.out(4), potential(3NEMO), bodytrans(3NEMO))
.SH AUTHOR
Eric Roberts (original SunOS version), Peter Teuben (sun4, coff, dl)
.SH FILES
.nf
.ta +1.5i
~/src/kernel/loadobj         loadobj.c, loadobj*.c
.fi
Much like \fIyapp(3NEMO)\fP, different implementations exist for
different operating systems.
.SH UPDATE HISTORY
.nf
.ta +1i +4i
23-jul-90       created         PJT
6-nov-91	slight doc improvements; loadobjNEXT.c   	PJT
.fi


src/kernel/loadobj/loadobj.c
/* loadobj.c:  
 *	Merely includes the right loadobjXXX.c for a particular machine.
 *	The Makefile defaults compilation with -D$(MACH) 
 *	The first few #ifdef's below are ''upper'' case, and can be used
 *	on ''any'' machine, typically using a -D... compile switch,
 *      If those fail in the test, it will try one of the
 *	machine supplied triggers, the order of e.g. sun after sparc is
 *	in this respect important. Then, if those fail too, the
 *	STUB loader is used. This one can also be loaded
 *	specifically with -DSTUB.
 *
 *	At the end also includes the testloadobj.c code for TESTBED
 *	testing (-DTESTBED)
 *
 *	28-oct-90	finalized for new Nemo				PJT
 *	18-nov-90	added GNU dld					PJT
 *	25-may-91 	added the STUB option - more doc		PJT 
 *       5-nov-91       added NEXT option                               PJT
 *      24-nov-91       added AIX for R6000 (IBM)			PJT
 *	28-nov-91	formally added MIPS (note NEXT/AIX/MIPS all bad)PJT
 *	20-jan-94	NULL is now STUB				pjt
 *	22-oct-96	DL is default for linux				pjt
 *      12-jul-03       DL is default for darwin (though needs dlcompat) PJT
 */

#include <stdinc.h>		/* standard Nemo include file */


/* patch up for some predefined defaults we understand: */

#if !defined(AUTO)
# if defined(sun) && defined(sparc) && defined(SYSV)
#  define DL 1
# elif defined(sgi) || defined(alpha) || defined(__alpha) 
#  define DL 1
# elif defined(linux) && defined(SYSV) || defined(darwin)
#  define DL 1
# endif
#endif

#if defined(DL)
char *loadobj_version="loadobjDL";
#include "loadobjDL.c"		/* new standard dlopen(3x) functions */

#elif defined(DLD)
char *loadobj_version="loadobjDLD";
#include "loadobjDLD.c"		/* GNU dld library */

#elif defined(XDL)
char *loadobj_version="loadobjXDL";
#include "loadobjXDL.c"		/* xdl (sun/dec) library */

#elif defined(STUB) || defined(sgi)
char *loadobj_version="loadobjSTUB";
#include "loadobjSTUB.c"	/* stubbed; gets you going on an odd machine */

#elif defined(sparc) && !defined(SYSV)
char *loadobj_version="loadobjSPARC";
#include "loadobjSPARC.c"	/* sun4 SPARC (works 99% of the time) */

#elif defined(sun) && !defined(SYSV)
char *loadobj_version="loadobjSUN";
#include "loadobjSUN.c"		/* general bsd (sun3), multiflow (?) */

#elif defined(_trace_)
char *loadobj_version="loadobj_mf";
#include "loadobj_mf.c"		/* MF hacking: does not work*/

#elif defined(COFF)
char *loadobj_version="loadobjCOFF";
#include "loadobjCOFF.c"	/* general sys5, not always works?? */

#elif defined(alliant)
char *loadobj_version="loadobjFX";
#include "loadobjFX.c"		/* alliant: does not work */

#elif defined(mc68k)
char *loadobj_version="loadobj3B1";
#include "loadobj3B1.c"		/* 3b1: doesn't always work - like COFF */

#elif defined(NeXT) || defined(NEXT)
char *loadobj_version="loadobjNEXT";
#include "loadobjNEXT.c"	/* NeXT: MACH O - doesn't work yet either */

#elif defined(aix)
char *loadobj_version="loadobjAIX";
#include "loadobjAIX.c"		/* IBM Risc - testing nov 91 */

#elif defined(MIPS) || defined(mips)
char *loadobj_version="loadobjMIPS";
#include "loadobjMIPS.c"	/* DEC5000 - testing nov 91 */

#elif defined(linux)
char *loadobj_version="loadobjLINUX";
#include "loadobjLINUX.c"		/* Linux - testing */

#else
				/* if all else failed....*/
char *loadobj_version="loadobjSTUB";
#include "loadobjSTUB.c"	/* stubbed; gets you going on an odd machine */

#endif

#if defined(TESTBED)
# include "testloadobj.c"
#endif
